spec/jwt_spec.rb

```diff
@@ -1,434 +1,502 @@
+      'ES256' => '',
+      'ES384' => '',
+      'ES512' => ''
+    }
+  end
+
+  after(:each) do
+    expect(OpenSSL.errors).to be_empty
   end
 
-  it 'encodes and decodes JWTs' do
-    secret = 'secret'
-    jwt = JWT.encode(@payload, secret)
-    decoded_payload = JWT.decode(jwt, secret)
-    expect(decoded_payload).to include(@payload)
-  end
+  context 'alg: NONE' do
+    let :alg do
+      'none'
+    end
 
-  it 'encodes and decodes JWTs for RSA signatures' do
-    private_key = OpenSSL::PKey::RSA.generate(512)
-    jwt = JWT.encode(@payload, private_key, 'RS256')
-    decoded_payload = JWT.decode(jwt, private_key.public_key)
-    expect(decoded_payload).to include(@payload)
-  end
+    it 'should generate a valid token' do
+      token = JWT.encode payload, nil, alg
 
-  it 'encodes and decodes JWTs for ECDSA P-256 signatures' do
-    private_key = OpenSSL::PKey::EC.new('prime256v1')
-    private_key.generate_key
-    public_key = OpenSSL::PKey::EC.new(private_key)
-    public_key.private_key = nil
-    jwt = JWT.encode(@payload, private_key, 'ES256')
-    decoded_payload = JWT.decode(jwt, public_key)
-    expect(decoded_payload).to include(@payload)
-  end
+      expect(token).to eq data['NONE']
+    end
 
-  it 'encodes and decodes JWTs for ECDSA P-384 signatures' do
-    private_key = OpenSSL::PKey::EC.new('secp384r1')
-    private_key.generate_key
-    public_key = OpenSSL::PKey::EC.new(private_key)
-    public_key.private_key = nil
-    jwt = JWT.encode(@payload, private_key, 'ES384')
-    decoded_payload = JWT.decode(jwt, public_key)
-    expect(decoded_payload).to include(@payload)
-  end
+    it 'should decode a valid token' do
+      payload, header = JWT.decode data['NONE'], nil, false
 
-  it 'encodes and decodes JWTs for ECDSA P-521 signatures' do
-    private_key = OpenSSL::PKey::EC.new('secp521r1')
-    private_key.generate_key
-    public_key = OpenSSL::PKey::EC.new(private_key)
-    public_key.private_key = nil
-    jwt = JWT.encode(@payload, private_key, 'ES512')
-    decoded_payload = JWT.decode(jwt, public_key)
-    expect(decoded_payload).to include(@payload)
+      expect(header['alg']).to eq alg
+      expect(payload).to eq payload
+    end
   end
 
-  it 'encodes and decodes JWTs with custom header fields' do
-    private_key = OpenSSL::PKey::RSA.generate(512)
-    jwt = JWT.encode(@payload, private_key, 'RS256', {'kid' => 'default'})
-    decoded_payload = JWT.decode(jwt) do |header|
-      expect(header['kid']).to eq('default')
-      private_key.public_key
+  %w(HS256 HS384 HS512).each do |alg|
+    context "alg: #{alg}" do
+      it 'should generate a valid token' do
+        token = JWT.encode payload, data[:secret], alg
+
+        expect(token).to eq data[alg]
+      end
+
+      it 'should decode a valid token' do
+        payload, header = JWT.decode data[alg], data[:secret]
+
+        expect(header['alg']).to eq alg
+        expect(payload).to eq payload
```

Comment:
payload will always eq payload... comparing the same objects.

I think you meant to compare the payload from the let block with the payload declared a couple of lines above, but because they have the same name, badness will ensue.

This looks to be repeated further down in the code too